---
title: "001.cold_acclimation_gene_expression"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(tidyverse)
library(ggdendro)
library(viridis)
library(cowplot)
library(wesanderson)
```


```{r data}
# RNA-seq samples
samples <- read_delim('/groups/nordborg/projects/cold_adaptation_16Cvs6C/003.transcriptome/001.8accessions/000.general_data/001.data/samples.txt', delim = '\t')

# variance stabilised transcript counts
counts.varStab <- read_csv('/groups/nordborg/projects/cold_adaptation_16Cvs6C/003.transcriptome/001.8accessions/001.transcript_quantification/003.results/transcriptCounts_varianceStabilised.csv')

# worldclim data
worldclim <- read_csv('/groups/nordborg/user/pieter.clauw/Documents/Source/Accessions/2029_modified_MN_SH_wc2.0_30s_bilinear.csv') %>%
  mutate(accession = as.factor(accession))

## cold acclimation genes

# CBF regulon according Park et al., 2015
CBF_UP.Park <- read_delim('/groups/nordborg/projects/cold_adaptation_16Cvs6C/003.transcriptome/001.8accessions/004.cold_acclimation_gene_expression/001.data/CBF_regulon_UP_ParkEtAl2015.txt', delim = '\t', col_names = 'gene')
CBF_DOWN.Park <- read_delim('/groups/nordborg/projects/cold_adaptation_16Cvs6C/003.transcriptome/001.8accessions/004.cold_acclimation_gene_expression/001.data/CBF_regulon_DOWN_ParkEtAl2015.txt', delim = '\t', col_names = 'gene')

# HSFC1 regulon according Park et al., 2015
HSFC1 <- read_delim('/groups/nordborg/projects/cold_adaptation_16Cvs6C/003.transcriptome/001.8accessions/004.cold_acclimation_gene_expression/001.data/HSFC1_regulon_ParkEtAl2015.txt', delim = '\t', col_names = 'gene')

# ZAT12 regulon according Vogel et al., 2005
ZAT12_UP <- read_csv('/groups/nordborg/projects/cold_adaptation_16Cvs6C/003.transcriptome/001.8accessions/004.cold_acclimation_gene_expression/001.data/ZAT12_upregulated_table9.csv') %>%
  filter(coldResponsive == 'x') %>%
  mutate(AGI = toupper(AGI))
ZAT12_DOWN <- read_csv('/groups/nordborg/projects/cold_adaptation_16Cvs6C/003.transcriptome/001.8accessions/004.cold_acclimation_gene_expression/001.data/ZAT12_downregulated_table10.csv') %>%
  filter(coldResponsive == 'x')%>%
  mutate(AGI = toupper(AGI))

# CBF and CAMTA transcription factors
CBF.TF <- c('AT4G25490', 'AT4G25470', 'AT4G25480', 'AT5G51990')
CAMTA.TF <- c('AT5G09410', 'AT5G64220', 'AT2G22300')

```
```{r functions}
# heatmap data matrix
heatmap.matrix <- function(data.tbl, rows, columns, values)
{
  data.tbl.wide <- data.tbl %>%
    dplyr::select(!!rows, !!columns, !!values) %>%
  pivot_wider(names_from = !!columns, values_from = !!values)
  data.tbl.mtrx <- as.matrix(data.tbl.wide[, -1])
  rownames(data.tbl.mtrx) <- data.tbl.wide[[rows]]
  return(data.tbl.mtrx)
}

# row dendrograms
heatmap.dendro.row <- function(heatmap.mtrx, k, dist.method = 'euclidean', clust.method = 'ward.D2')
{
  clust.row <- hclust(dist(heatmap.mtrx, method = dist.method), method = clust.method)
  dd.row <- as.dendrogram(clust.row)
  row.ord <- order.dendrogram(dd.row)
  ordered_row_names <- row.names(heatmap.mtrx[row.ord, ])
  dendro_data_row <- dendro_data(dd.row, type = "rectangle")
  
  gene.clust <- cutree(clust.row, k = k)
  clust.df <- data.frame(label=names(gene.clust), cluster=factor(gene.clust))
  dendro_data_row[["labels"]]   <- merge(dendro_data_row[["labels"]],clust.df, by="label")
  rect <- aggregate(x~cluster,label(dendro_data_row),range)
  rect <- data.frame(rect$cluster,rect$x)
  rect$mid <- rowMeans(rect[, c(2, 3)])
  ymax <- mean(clust.row$height[length(clust.row$height)-((k-2):(k-1))])
  
  return(list('ordered_row_names' = ordered_row_names,
              'dendro_data_row' = dendro_data_row,
              'gene.clust' = gene.clust,
              'rect' = rect,
              'ymax' = ymax))
}

# col dendrograms
heatmap.dendro.col <- function(heatmap.mtrx, dist.method = 'euclidean', clust.method = 'ward.D2')
{
  clust.col <- hclust(dist(t(heatmap.mtrx), method = dist.method), method = clust.method)
  dd.col <- as.dendrogram(clust.col)
  col.ord <- order.dendrogram(dd.col)
  ordered_col_names <- colnames(heatmap.mtrx[, col.ord])
  dendro_data_col <- dendro_data(dd.col, type = "rectangle")
  return(list('ordered_col_names' = ordered_col_names,
              'dendro_data_col' = dendro_data_col))
}

# GO enrichment
GOenrich <- function(genesInterest, geneUniverse, onto = 'BP')
{
  # GO enrichment - Biological Processes
  # specifiy gene list within gene universie
  #geneUniverse <- mcols(araport11.genes.GR)$gene
  geneList <- factor(as.integer(geneUniverse %in% genesInterest))
  names(geneList) <- geneUniverse

  # build GO dataset with TAIR annotation
  GO.dat <- new('topGOdata',
                description = 'gwas results GO',
                ontology = onto,
                allGenes = geneList,
                annot = annFUN.org,
                mapping = 'org.At.tair.db',
                nodeSize = 5)

  resultClassic <- runTest(GO.dat, algorithm = "classic", statistic = "fisher")
  resultWeight01 <- runTest(GO.dat, algorithm = "weight01", statistic = "fisher")
  resultElim <- runTest(GO.dat, algorithm = "elim", statistic = "fisher")
  resultWeight <- runTest(GO.dat, algorithm = "weight", statistic = "fisher")
  
  GOresults <- list(classic = resultClassic, weight01 = resultWeight01, elim = resultElim, weight = resultWeight)
  
  # gather data
  allRes <- GenTable(GO.dat,
                     classic = GOresults$classic,
                     weight01 = GOresults$weight01,
                     elim = GOresults$elim,
                     weight = GOresults$weight,
                     orderBy = 'classic',
                     ranksOf = 'weight01',
                     topNodes = length(usedGO(GO.dat)))
  
  # filter for significance
  allRes.filt <- as_tibble(allRes) %>%
    mutate(classic = as.numeric(classic),
           weight01 = as.numeric(weight01),
           elim = as.numeric(elim),
           weight = as.numeric(weight)) %>%
    filter(classic <= 0.1 | weight01 <= 0.1 | elim <= 0.1 | weight <= 0.1)
  
  return(allRes.filt)
}
```

```{r filter samples}
# filter only sequenced samples
samples <- filter(samples, Selected == 'yes') %>%
  mutate(accession = as.factor(accession),
    temperature = as.factor(temperature),
         replicate = as.factor(replicate),
         experiment = as.factor(paste(temperature, replicate, sep = '_')),
         sampleName = paste(accession, temperature, replicate, sep = '_')) %>%
    left_join(., worldclim[, c('accession', 'bio_11')], by = 'accession')
```

```{r heatmap of CBF}
# prepare data
counts.varstab.CBF <- counts.varStab %>%
  pivot_longer(contains('AT', ignore.case = F), names_to = 'gene', values_to = 'expression') %>%
  mutate(accession = as.factor(accession)) %>%
  filter(gene %in% c(CBF_UP.Park$gene, CBF_DOWN.Park$gene, CBF.TF, ZAT12_DOWN$AGI, ZAT12_UP$AGI, HSFC1$gene)) %>%
  dplyr::select(gene, sample, accession, temperature, expression) %>%
  group_by(accession, temperature, gene) %>%
  summarise(mean(expression), .groups = 'drop') %>%
  mutate(acnTemp = paste(accession, temperature, sep = '_'), expression = `mean(expression)`) %>%
  dplyr::select(-`mean(expression)`) %>%
  group_by(gene) %>%
  mutate(expression = scale(expression))
  
# Convert to a matrix for clustering
counts.varstab.CBF.wide <- counts.varstab.CBF %>%
  dplyr::select(gene, acnTemp, expression) %>%
  pivot_wider(id_cols = gene, names_from = acnTemp, values_from = expression)

counts.varstab.CBF.mtrx <- as.matrix(counts.varstab.CBF.wide[, -1])
rownames(counts.varstab.CBF.mtrx) <- counts.varstab.CBF.wide$gene

# climate annotation
clim.annot <- mutate(samples, acnTemp = paste(accession, temperature, sep = '_')) %>%
  dplyr::select(acnTemp, bio_11) %>%
  distinct()
clim.annot <-data.frame(bio_11 = clim.annot$bio_11, row.names = clim.annot$acnTemp)


counts.varstab.CBF.long <- pivot_longer(counts.varstab.CBF.wide, values_to = 'expression', names_to = 'acn_temp', cols = -gene)

# Cluster rows
k <- 7
clust.row <- hclust(dist(counts.varstab.CBF.mtrx, method = 'euclidean'), method = 'ward.D2')
dd.row <- as.dendrogram(clust.row)
row.ord <- order.dendrogram(dd.row)
ordered_row_names <- row.names(counts.varstab.CBF.mtrx[row.ord, ])
counts.varstab.CBF.long$gene <- factor(counts.varstab.CBF.long$gene, levels = ordered_row_names)
dendro_data_row <- dendro_data(dd.row, type = "rectangle")

gene.clust <- cutree(clust.row, k = k)
clust.df <- data.frame(label=names(gene.clust), cluster=factor(gene.clust))
clust.df$label <- factor(clust.df$label, levels = ordered_row_names, ordered = T)
clust.df <- clust.df[order(clust.df$label, decreasing = T), ]
clust.df$cluster.name <- as.numeric(factor(clust.df$cluster, levels = unique(clust.df$cluster)))
dendro_data_row[["labels"]]   <- merge(dendro_data_row[["labels"]],clust.df, by="label")
rect <- aggregate(x~cluster.name,label(dendro_data_row),range)
rect <- data.frame(rect$cluster.name,rect$x)
rect$mid <- rowMeans(rect[, c(2, 3)])
ymax <- mean(clust.row$height[length(clust.row$height)-((k-2):(k-1))])

    
# Cluster columns
clust.col <- hclust(dist(t(counts.varstab.CBF.mtrx), method = 'euclidean'), method = 'ward.D2')
dd.col <- as.dendrogram(clust.col)
col.ord <- order.dendrogram(dd.col)
ordered_col_names <- colnames(counts.varstab.CBF.mtrx[, col.ord])
counts.varstab.CBF.long$acn_temp <- factor(counts.varstab.CBF.long$acn_temp, levels = ordered_col_names)

heat_plot <- ggplot(counts.varstab.CBF.long, aes(x = acn_temp, y = gene, fill = expression)) + 
  geom_tile() +
  xlab("") + 
	ylab("") +
  labs(fill = 'gene-wide\nstandardized\nexpression') +
	theme_minimal() + 
  theme(text = element_text(size = 10),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5),
        axis.text.y = element_blank(),
        axis.line  = element_blank(),
        axis.ticks = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank()) +
  scale_fill_viridis(discrete = F)

dendro_row_box <- axis_canvas(heat_plot, axis = "y", coord_flip = TRUE) + 
  geom_segment(data = segment(dendro_data_row), aes(y = -y, x = x, xend = xend, yend = -yend), size = 0.2) +
  geom_rect(data = rect, aes(xmin=X1-0.3, xmax=X2+0.3, ymin = -ymax - 2, ymax = 0), 
            color=wes_palette('Darjeeling1')[1], fill=NA, size = 0.1) +
  geom_text(data = rect, aes(x = mid, y = -ymax - 2.7 , label = rect.cluster.name), col = wes_palette('Darjeeling1')[1]) +
  coord_flip() +
  theme_dendro()

dendro_row <- axis_canvas(heat_plot, axis = "y", coord_flip = T) + 
  geom_segment(data = segment(dendro_data_row), aes(y = -y, x = x, xend = xend, yend = -yend), size = 0.2) +
  coord_flip() +
  theme_dendro()
  

# annotate bio11
bio11.annot <- tibble(acnTemp = ordered_col_names) %>%
  separate(acnTemp, c('accession', NA), remove = F) %>%
  left_join(., dplyr::select(worldclim, accession, bio_11), by = 'accession') %>%
  mutate(sampleNr = c(1:length(acnTemp))) %>%
  dplyr::select(-accession)

bio11_col_annot <- axis_canvas(heat_plot, axis = 'x', data = bio11.annot, mapping = aes(x = sampleNr)) +
  geom_tile(aes(y = 1, fill = bio_11)) +
  theme(legend.position = 'right',
        legend.direction = 'horizontal',
        legend.key.height = unit(0.5, 'lines'),
        legend.key.width = unit(1, 'lines'),
        legend.title = element_blank())
bio11_col_annot.lgnd <- cowplot::get_legend(bio11_col_annot)

dendro_data_col <- dendro_data(dd.col, type = "rectangle")
dendro_col <- axis_canvas(heat_plot, axis = "x") + 
  geom_segment(data = segment(dendro_data_col), aes(x = x, y = y, xend = xend, yend = yend), size = 0.2)

# plot heatmap
final_plot <- heat_plot

# create empty plot for heatmap/ climate annotation separation
pnull <- ggdraw()

# add sample dendrogram and climayte annotation to heatmap
final_plot <- 
  insert_xaxis_grob(
    insert_xaxis_grob(
      insert_xaxis_grob(
        final_plot, dendro_col, unit(0.2, "null"), position = "top"),
      bio11_col_annot, unit(0.02, "null"), position = 'top'),
    pnull, unit(0.01, 'null'), position = 'top')

# add gene dendrogram  with clusters to heatmap
final_plot <- insert_yaxis_grob(final_plot, dendro_row_box, unit(0.2, "null"), position = "left")

p <- ggdraw(final_plot) +
  draw_grob(bio11_col_annot.lgnd, scale = 0.5, x = 0.42, y = 0.31) +
  draw_text('temp. coldest quart.', x = 0.92, y = 0.89, size = 10)

ggsave(filename = '/groups/nordborg/projects/cold_adaptation_16Cvs6C/003.transcriptome/001.8accessions/004.cold_acclimation_gene_expression/004.plots/heatmap_cold_acclimation.pdf',
       plot = p)

print(p)

CBF.gene.clusters <- map(c(1:max(clust.df$cluster.name)),                          function(i){clust.df$label[clust.df$cluster.name == i]})
```