---
title: "visualise gene expression"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(tidyverse)
```

## intro
script to visualize expression differences between temperatures and accessions for specific gene sets.
Expression data is based on RNA-Seq data of 8 accessions grown in either 16ºC or 6ªC.


### read in data
```{r data}
# expression data of 19055 genes for 48 samples.
expression.tbl <- read_csv(file = '/groups/nordborg/projects/cold_adaptation_16Cvs6C/transcriptome/8accessions/transcript_quantification/transcriptCounts_varianceStabilised.csv')

# gene list.
# Simple .csv file with one column containing the genes of interest. Genes specified as ATG codes
genes <- read_csv(file = '/groups/nordborg/projects/cold_adaptation_16Cvs6C/transcriptome/8accessions/general_data/test_gene_set.csv') %>%
  # take out the first column and save as a vector
  pull(gene)
```
## make plots
```{r plot expression per accession}
# start from the data in 'expression'
expression.tbl %>%
  # select required columns. any_of(genes) selects the columns of the genes specified in 'genes'
  select(accession, temperature, replicate, any_of(genes)) %>%
  # change 'accession' from numeric to a discrete factor
  mutate(accession = as.factor(accession)) %>%
  # transform data to long format
  pivot_longer(cols = matches('AT.G'), names_to = 'gene', values_to = 'expression') %>%
  # start building the plot
  # x-axis is accession, y-axis the expression values, data of each gene will get a different color
  ggplot(aes(x = accession, y = expression, color = gene)) +
  # specify that data should be plotted as points. each replicate gets a different shape.
  geom_point(aes(shape = replicate)) +
  # specify that data should also be plotted as lines. each temperature gets a different type of line. data get grouped by temperature, replicate and gene
  geom_line(aes(linetype = temperature, group = interaction(temperature, replicate, gene)))
```

```{r plot expression per temperature}
expression.tbl %>%
  dplyr::select(accession, temperature, replicate, any_of(genes)) %>%
  mutate(accession = as.factor(accession)) %>%
  pivot_longer(cols = matches('AT.G'), names_to = 'gene', values_to = 'expression') %>%
  # x-axis is temperature, y-axis the expression values, data of each gene will get a different color
  ggplot(aes(x = temperature, y = expression, color = gene)) +
  geom_point(aes(shape = replicate)) +
  # specify that data should also be plotted as lines. data get grouped by temperature, replicate and gene
  geom_line(aes(group = interaction(accession, replicate, gene)))
```

```{r plot average expression per gene per temperature per accession}
expression.tbl %>%
  dplyr::select(accession, temperature, replicate, any_of(genes)) %>%
  mutate(accession = as.factor(accession)) %>%
  pivot_longer(cols = matches('AT.G'), names_to = 'gene', values_to = 'expression') %>%
  # group data by gene, accession and temperature. Future processing will be done for each group independently
  group_by(gene, accession, temperature) %>%
  # specify 'expression.avg' as median expression for each of the groups. In our case the median over the different replicates for each gene for each accession in each temperature. Drop the grouping structure when done.
  summarise(expression.avg = median(expression), .groups = 'drop') %>%
  ggplot(aes(x = temperature, y = expression.avg, color = gene)) +
  geom_point(aes(shape = accession)) +
  geom_line(aes(group = interaction(gene, accession))) +
  # specify the different point shapes manually. By default there are not more than 6 shapes used.
  scale_shape_manual(values=c(15, 19, 17, 18, 4, 8, 9, 11))
```

```{r plot average expression per gene per temperature}
expression.tbl %>%
  dplyr::select(accession, temperature, replicate, any_of(genes)) %>%
  mutate(accession = as.factor(accession)) %>%
  # transform data to long format
  pivot_longer(cols = matches('AT.G'), names_to = 'gene', values_to = 'expression') %>%
  group_by(gene, temperature) %>%
  summarise(expression.avg = median(expression), .groups = 'drop') %>%
  ggplot(aes(x = temperature, y = expression.avg, color = gene)) +
  geom_point() +
  geom_line(aes(group = interaction(gene))) +
  scale_shape_manual(values=c(15, 19, 17, 18, 4, 8, 9, 11))
```


