---
title: "network_temperature_specific"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(tidyverse)
library(WGCNA)
library(tximport)
library(DESeq2)
library(biomaRt)
library(ggpubr)
```

```{r data}
# gene description
ensembl <- useMart(biomart="plants_mart", host="plants.ensembl.org", dataset="athaliana_eg_gene")

# expression estimaes - variance stabilised
expression <- read_csv('/groups/nordborg/projects/cold_adaptation_16Cvs6C/003.transcriptome/001.8accessions/001.transcript_quantification/003.results/transcriptCounts_varianceStabilised.csv') %>%
  mutate(accession = as.factor(accession))

# gene categories Sonia
gene_cat <- read_delim('/groups/nordborg/user/pieter.clauw/Documents/Experiments/UltimateQandD/Results/Transcriptome/diffExpr_Sonia/Category_genes.txt', delim = '\t')

# output directory
out.dir <- '/groups/nordborg/projects/cold_adaptation_16Cvs6C/003.transcriptome/001.8accessions/003.coExpression/003.results/'

# specify temperatures
temperatures <- c('16C', '6C')

```
```{r functions}
# filter genes with too many missing data and/or zero variance
geneFilter <- function(exprData)
{
  gsg <- goodSamplesGenes(exprData, verbose = 3)
  if (!gsg$allOK)
  {
    # Optionally, print the gene and sample names that were removed:
    if (sum(!gsg$goodGenes)>0)
    {printFlush(paste("Removing genes:", paste(names(exprData)[!gsg$goodGenes], collapse = ", ")))}
    if (sum(!gsg$goodSamples)>0)
    {printFlush(paste("Removing samples:", paste(rownames(exprData)[!gsg$goodSamples], collapse = ", ")))}
    # Remove the offending genes and samples from the data:
    return(exprData[gsg$goodSamples, gsg$goodGenes])
  }
  else
  {return(exprData)}
}

# plot the scale-free topology fit index as functio of soft threshold power
plotSoftThresholdChoices <- function(exprData, minSofthThreshold = 12, maxSoftThreshold = 20, steps = 2)
{
  # Choose a set of soft-thresholding powers
  powers <- c(c(1:10), seq(from = minSofthThreshold, to = maxSoftThreshold, by = steps))
  # Call the network topology analysis function
  sft <- pickSoftThreshold(exprData, powerVector = powers, verbose = 5)
  # Plot the results:
  sizeGrWindow(9, 5)
  par(mfrow = c(1,2));
  cex1 = 0.9;
  # Scale-free topology fit index as a function of the soft-thresholding power
  plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
       main = paste("Scale independence"));
  text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       labels=powers,cex=cex1,col="red");
  # this line corresponds to using an R^2 cut-off of h
  abline(h=0.90,col="red")
  # Mean connectivity as a function of the soft-thresholding power
  plot(sft$fitIndices[,1], sft$fitIndices[,5],
       xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
       main = paste("Mean connectivity"))
  text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
}
```

```{r split both temperatures}
expression.lst <- map(temperatures, function(temp){
  expression %>%
    filter(temperature == !!temp)})  %>%
  set_names(temperatures)

```

## Network for each temperature
```{r filter genes}
expression.mtrx <- map(temperatures, function(temp){
  expression.lst[[temp]] %>%
    dplyr::select(matches('AT.G.{5}')) %>%
    as.matrix()}) %>%
  set_names(temperatures)

# gene filtering
expression.mtrx.filtered <- map(temperatures, function(temp){
  # filter out genes with expression data for less than 1/2 samples
  # and filter out genes with variance == 0
  filt <- geneFilter(expression.mtrx[[temp]])
  
  # select only genes with differential expression (genotype effect or temperature effect or GxE)
  filt <- filt[, gene_cat$genes[which(gene_cat$genes %in% colnames(filt))]]
  rownames(filt) <- pull(expression.lst[[temp]], sample)
  return(filt)
  
}) %>%
  set_names(temperatures)
```
```{r PCA on subset of genes}
map(temperatures, function(temp){
  pcaData<- prcomp(expression.mtrx.filtered[[temp]])
  percentage <- round(pcaData$sdev / sum(pcaData$sdev) * 100, 2)
  percentage <- paste(colnames(pcaData$x), "(", paste( as.character(percentage), "%", ")", sep="") )
  
  as_tibble(pcaData$x, rownames = 'sample') %>%
    left_join(., dplyr::select(expression.lst[[temp]], sample, accession, temperature, replicate), by = 'sample') %>%
    ggplot(aes(PC1, PC2, color = accession, shape = replicate)) +
    geom_point(size = 3) +
    labs(x = percentage[1],
         y = percentage[2],
         title = temp)
})

```

```{r make list framework for WGCNA analysis}
wgcna.data <- list()
for (temp in temperatures){
 wgcna.data[[temp]][['expression']] <- expression.mtrx.filtered[[temp]] 
}
```

```{r test soft threshold 6ºC}
#softthrsholdTest <- map(temperatures, function(temp){
#  return(plotSoftThresholdChoices(wgcna.data[[temp]]$expression, maxSoftThreshold = 20))  
#})



#for(temp in temperatures){
#  plotSoftThresholdChoices(wgcna.data[[temp]]$expression, maxSoftThreshold = 20)
#}


plotSoftThresholdChoices(wgcna.data[['6C']]$expression, maxSoftThreshold = 20)

```

```{r test soft threshold 16ºC}
plotSoftThresholdChoices(wgcna.data[['16C']]$expression, maxSoftThreshold = 20)
```

A soft threhsold of 7 was picked for boht temnperatures
This soft threshold has a scale free topology model right above  0.8 and still a mean connectivity of above 0.95


```{r choose softThreshold}
for (temp in temperatures)
{
  wgcna.data[[temp]]$softThreshold <- 7
}
```

```{r adjacencies}
for (temp in temperatures)
{
  wgcna.data[[temp]]$adjacency <- abs(bicor(wgcna.data[[temp]]$expression, use = "p", maxPOutliers = 0.10))^wgcna.data[[temp]]$softThreshold
}
```

```{r TOM}
for (temp in temperatures)
{
  wgcna.data[[temp]]$TOM <- TOMsimilarity(wgcna.data[[temp]]$adjacency)  
}

# prevent too much memory usage by not saving the dissimilarity matrix but create it when needed with 1-TOM
#expr.filter[[i]]$dissTOM <- 1 - expr.filter[[i]]$TOM
```
```{r dendrogram 6ºC}
collectGarbage()
wgcna.data[['6C']]$geneTree <- hclust(as.dist(1- wgcna.data[['6C']]$TOM), method = "average")
  sizeGrWindow(12,9)
  x <- plot(wgcna.data[['6C']]$geneTree, xlab="", sub="", main = "Gene clustering on TOM-based dissimilarity - 6ºC", labels = FALSE, hang = 0.04)
```

```{r dendrogram 16ºC}
collectGarbage()
wgcna.data[['16C']]$geneTree <- hclust(as.dist(1- wgcna.data[['16C']]$TOM), method = "average")
  sizeGrWindow(12,9)
  x <- plot(wgcna.data[['16C']]$geneTree, xlab="", sub="", main = "Gene clustering on TOM-based dissimilarity - 16ºC", labels = FALSE, hang = 0.04)
```

```{r get modules}
minModuleSize <- 10
for (temp in temperatures)
{
  wgcna.data[[temp]]$dynamicMods <- cutreeDynamic(dendro = wgcna.data[[temp]]$geneTree, distM = 1 - wgcna.data[[temp]]$TOM, deepSplit = 1, pamRespectsDendro = FALSE, minClusterSize = minModuleSize)
}

table(wgcna.data[['6C']]$dynamicMods)
table(wgcna.data[['16C']]$dynamicMods)
```


```{r size distribution unmerged}
hist.lst <- map(temperatures, function(temp){
  tibble('moduleSize' = table(wgcna.data[[temp]]$dynamicMods)) %>%
  ggplot(aes(x = moduleSize)) +
    geom_histogram() +
    ggtitle(temp)
})

ggarrange(plotlist = hist.lst)
```
```{r plot modules 6C}
temp <- '6C'
wgcna.data[[temp]]$dynamicColors <- labels2colors(wgcna.data[[temp]]$dynamicMods)
sizeGrWindow(8,6)
plotDendroAndColors(wgcna.data[[temp]]$geneTree, wgcna.data[[temp]]$dynamicColors, "Dynamic Tree Cut", dendroLabels = FALSE, hang = 0.03, addGuide = TRUE, guideHang = 0.05, main = paste("Gene dendrogram and module colors", temp, sep = ' - '))
```
```{r plot modules 16C}
temp <- '16C'
wgcna.data[[temp]]$dynamicColors <- labels2colors(wgcna.data[[temp]]$dynamicMods)
sizeGrWindow(8,6)
plotDendroAndColors(wgcna.data[[temp]]$geneTree, wgcna.data[[temp]]$dynamicColors, "Dynamic Tree Cut", dendroLabels = FALSE, hang = 0.03, addGuide = TRUE, guideHang = 0.05, main = paste("Gene dendrogram and module colors", temp, sep = ' - '))
```
```{r decide level to merge - 6C}
MEDissThres <- 0.10
temp <- '6C'
  ###calculate eigengenes
  MEList <- moduleEigengenes(wgcna.data[[temp]]$expression, colors=  wgcna.data[[temp]]$dynamicMods, softPower = wgcna.data[[temp]]$softThreshold)
  MEs <- MEList$eigengenes
  # Calculate dissimilarity of module eigengenes
  MEDiss = 1-cor(MEs)
  # Cluster module eigengenes
  METree = hclust(as.dist(MEDiss), method = "average")
  # Plot the result
  sizeGrWindow(7, 6)
  plot(METree, main = paste("Clustering of module eigengenes", temp, sep = ' - '), xlab = "", sub = "", cex=0.3)
  # Plot the cut line into the dendrogram
  abline(h=MEDissThres, col = "red")
```
```{r decide level to merge - 16C}
MEDissThres <- 0.10
temp <- '16C'
  ###calculate eigengenes
  MEList <- moduleEigengenes(wgcna.data[[temp]]$expression, colors=  wgcna.data[[temp]]$dynamicMods, softPower = wgcna.data[[temp]]$softThreshold)
  MEs <- MEList$eigengenes
  # Calculate dissimilarity of module eigengenes
  MEDiss = 1-cor(MEs)
  # Cluster module eigengenes
  METree = hclust(as.dist(MEDiss), method = "average")
  # Plot the result
  sizeGrWindow(7, 6)
  plot(METree, main = paste("Clustering of module eigengenes", temp, sep = ' - '), xlab = "", sub = "", cex=0.3)
  # Plot the cut line into the dendrogram
  abline(h=MEDissThres, col = "red")
```

```{r merge similar clusters}
#this step is different comparing to the tutorial for set-specific network construction and it's taken from the tutorial for consensus as it seems to work better (with the first one the moduleLabels it's not giving the same number of modules as the moduleColors, don't know why)
MEDissThres <- 0.10
for (temp in temperatures)
{
  merge <- mergeCloseModules(wgcna.data[[temp]]$expression, wgcna.data[[temp]]$dynamicMods, cutHeight = MEDissThres, verbose = 3)
  # The merged module labels
  wgcna.data[[temp]]$mergedLabels <- merge$colors
  # translate into module colors
  wgcna.data[[temp]]$mergedColors <- labels2colors(wgcna.data[[temp]]$mergedLabels)
  # Eigengenes of the new merged modules
  wgcna.data[[temp]]$MEs <- merge$newMEs
}
```
```{r plot merged clusters - 6C}
sizeGrWindow(12, 9)
plotDendroAndColors(wgcna.data[['6C']]$geneTree, cbind(wgcna.data[['6C']]$dynamicColors, wgcna.data[['6C']]$mergedColors), c("Dynamic Tree Cut", "Merged dynamic"), dendroLabels = FALSE, hang = 0.03, addGuide = TRUE, guideHang = 0.05, main = 'Modules after merging - 6ºC')
table(wgcna.data[['6C']]$mergedLabels)
length(unique(wgcna.data[['6C']]$mergedLabels))  

```

```{r plot merged clusters - 16C}
sizeGrWindow(12, 9)
plotDendroAndColors(wgcna.data[['16C']]$geneTree, cbind(wgcna.data[['16C']]$dynamicColors, wgcna.data[['16C']]$mergedColors), c("Dynamic Tree Cut", "Merged dynamic"), dendroLabels = FALSE, hang = 0.03, addGuide = TRUE, guideHang = 0.05, main = 'Modules after merging - 16ºC')
table(wgcna.data[['16C']]$mergedLabels)
length(unique(wgcna.data[['16C']]$mergedLabels)) 
```
```{r size distribution merged}
hist.lst <- map(temperatures, function(temp){
  tibble('moduleSize' = table(wgcna.data[[temp]]$mergedColors)) %>%
  ggplot(aes(x = moduleSize)) +
    geom_histogram() +
    ggtitle(temp)
})

ggarrange(plotlist = hist.lst)
```
```{r export network to cytoscape}
for (temp in temperatures)
{
  gene.names <- getBM(attributes = 'external_gene_name', filters = 'ensembl_gene_id', values = colnames(wgcna.data[[temp]]$expression), mart =ensembl)
  gene.attributes <- tibble(gene = colnames(wgcna.data[[temp]]$expression)) %>%
  left_join(., gene_cat, by = c('gene' = 'genes'))

  # define threshold
  trsh = 0.2

  # export network
  exportNetworkToCytoscape(wgcna.data[[temp]]$TOM,
                         edgeFile = paste0(out.dir, 'network_', temp, '_full_edge_', trsh, '.txt'),
                         nodeFile = paste0(out.dir, 'network_', temp, '_full_node_', trsh, '.txt'),
                         weighted = T,
                         threshold = trsh,
                         nodeNames = colnames(wgcna.data[[temp]]$expression),
                         altNodeNames = gene.names$external_gene_name,
                         nodeAttr = gene.attributes)
}
```

```{r save networks}
save(wgcna.data, file = paste(out.dir, 'network_temperature_specific.Rda'))
```

