---
title: "network_temperature_specific"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(tidyverse)
library(WGCNA)
library(tximport)
library(DESeq2)
library(biomaRt)
library(doParallel)
```

```{r data}
# gene description
ensembl <- useMart(biomart="plants_mart", host="plants.ensembl.org", dataset="athaliana_eg_gene")

# expression estimaes - variance stabilised
expression <- read_csv('/groups/nordborg/projects/cold_adaptation_16Cvs6C/003.transcriptome/001.8accessions/001.transcript_quantification/transcriptCounts_varianceStabilised.csv') %>%
  mutate(accession = as.factor(accession))

# gene categories Sonia
gene_cat <- read_delim('/groups/nordborg/user/pieter.clauw/Documents/Experiments/UltimateQandD/Results/Transcriptome/diffExpr_Sonia/Category_genes.txt', delim = '\t')

# output directory
out.dir <- '/groups/nordborg/projects/cold_adaptation_16Cvs6C/003.transcriptome/001.8accessions/003.coExpression/003.results/'

# specify temperatures
temperatures <- c('16C', '6C')

```
```{r functions}
# filter genes with too many missing data and/or zero variance
geneFilter <- function(exprData)
{
  gsg <- goodSamplesGenes(exprData, verbose = 3)
  if (!gsg$allOK)
  {
    # Optionally, print the gene and sample names that were removed:
    if (sum(!gsg$goodGenes)>0)
    {printFlush(paste("Removing genes:", paste(names(exprData)[!gsg$goodGenes], collapse = ", ")))}
    if (sum(!gsg$goodSamples)>0)
    {printFlush(paste("Removing samples:", paste(rownames(exprData)[!gsg$goodSamples], collapse = ", ")))}
    # Remove the offending genes and samples from the data:
    return(exprData[gsg$goodSamples, gsg$goodGenes])
  }
  else
  {return(exprData)}
}

# plot the scale-free topology fit index as functio of soft threshold power
plotSoftThresholdChoices <- function(exprData, minSofthThreshold = 12, maxSoftThreshold = 20, steps = 2)
{
  # Choose a set of soft-thresholding powers
  powers <- c(c(1:10), seq(from = minSofthThreshold, to = maxSoftThreshold, by = steps))
  # Call the network topology analysis function
  sft <- pickSoftThreshold(exprData, powerVector = powers, verbose = 5)
  # Plot the results:
  sizeGrWindow(9, 5)
  par(mfrow = c(1,2));
  cex1 = 0.9;
  # Scale-free topology fit index as a function of the soft-thresholding power
  plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
       main = paste("Scale independence"));
  text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       labels=powers,cex=cex1,col="red");
  # this line corresponds to using an R^2 cut-off of h
  abline(h=0.90,col="red")
  # Mean connectivity as a function of the soft-thresholding power
  plot(sft$fitIndices[,1], sft$fitIndices[,5],
       xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
       main = paste("Mean connectivity"))
  text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
}
```

```{r split both temperatures}
expression.lst <- map(temperatures, function(temp){
  expression %>%
    filter(temperature == !!temp)})  %>%
  set_names(temperatures)

```

## Network for each temperature
```{r filter genes}
expression.mtrx <- map(temperatures, function(temp){
  expression.lst[[temp]] %>%
    dplyr::select(matches('AT.G.{5}')) %>%
    as.matrix()}) %>%
  set_names(temperatures)

# gene filtering
expression.mtrx.filtered <- map(temperatures, function(temp){
  # filter out genes with expression data for less than 1/2 samples
  # and filter out genes with variance == 0
  filt <- geneFilter(expression.mtrx[[temp]])
  
  # select only genes with differential expression (genotype effect or temperature effect or GxE)
  filt <- filt[, gene_cat$genes[which(gene_cat$genes %in% colnames(filt))]]
  rownames(filt) <- pull(expression.lst[[temp]], sample)
  return(filt)
  
}) %>%
  set_names(temperatures)
```
```{r PCA on subset of genes}
map(temperatures, function(temp){
  pcaData<- prcomp(expression.mtrx.filtered[[temp]])
  percentage <- round(pcaData$sdev / sum(pcaData$sdev) * 100, 2)
  percentage <- paste(colnames(pcaData$x), "(", paste( as.character(percentage), "%", ")", sep="") )
  
  as_tibble(pcaData$x, rownames = 'sample') %>%
    left_join(., dplyr::select(expression.lst[[temp]], sample, accession, temperature, replicate), by = 'sample') %>%
    ggplot(aes(PC1, PC2, color = accession, shape = replicate)) +
    geom_point(size = 3) +
    labs(x = percentage[1],
         y = percentage[2],
         title = temp)
})

```

```{r make list framework for WGCNA analysis}
wgcna.data <- list()
for (temp in temperatures){
 wgcna.data[[temp]][['expression']] <- expression.mtrx.filtered[[temp]] 
}
```

```{r test soft threshold}

#TODO: try this
registerDoParallel()
softthrsholdTest <- map(temperatures, function(temp){
  return(plotSoftThresholdChoices(wgcna.data[[temp]]$expression, maxSoftThreshold = 20))  
})

for(temp in temperatures){
  plotSoftThresholdChoices(wgcna.data[[temp]]$expression, maxSoftThreshold = 20)
}


```








